firebase-firestore/src/test/java/com/google/firebase/firestore/local/BundleSerializerTest.java

```diff
@@ -0,0 +1,701 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.local;
+
+import static com.google.firebase.firestore.testutil.TestUtil.filter;
+import static com.google.firebase.firestore.testutil.TestUtil.key;
+import static com.google.firebase.firestore.testutil.TestUtil.orderBy;
+import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
+
+import com.google.firebase.firestore.core.Bound;
+import com.google.firebase.firestore.core.Query;
+import com.google.firebase.firestore.model.DatabaseId;
+import com.google.firebase.firestore.model.Document;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.ObjectValue;
+import com.google.firebase.firestore.model.ResourcePath;
+import com.google.firebase.firestore.model.SnapshotVersion;
+import com.google.firebase.firestore.remote.RemoteSerializer;
+import com.google.firebase.firestore.testutil.TestUtil;
+import com.google.firestore.v1.ArrayValue;
+import com.google.firestore.v1.MapValue;
+import com.google.firestore.v1.Value;
+import com.google.protobuf.NullValue;
+import com.google.protobuf.Timestamp;
+import com.google.type.LatLng;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.List;
+import java.util.TimeZone;
+import org.json.JSONException;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest = Config.NONE)
+public class BundleSerializerTest {
+  private static String TEST_PROJECT = "projects/project/databases/(default)/documents";
+  private static String TEST_DOCUMENT = TEST_PROJECT + "/coll/doc";
+
+  private BundleSerializer serializer =
+      new BundleSerializer(new RemoteSerializer(DatabaseId.forProject("project")));
+
+  // Value decoding tests
+
+  @Test
+  public void testDecodesNullValue() throws JSONException {
+    String json = "{ nullValue: null }";
+    Value proto = Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
+    assertValueRoundTrip(json, proto);
+  }
+
+  @Test
+  public void testDecodesBooleanValue() throws JSONException {
+    List<Boolean> tests = asList(true, false);
+    for (Boolean test : tests) {
+      String json = "{ booleanValue: " + test + " }";
+      Value proto = Value.newBuilder().setBooleanValue(test).build();
+      assertValueRoundTrip(json, proto);
+    }
+  }
+
+  @Test
+  public void testDecodesIntegerValues() throws JSONException {
+    List<Long> tests = asList(Long.MIN_VALUE, -100L, -1L, 0L, 1L, 100L, Long.MAX_VALUE);
+    for (Long test : tests) {
+      String json = "{ integerValue: '" + test + "' }";
```

Comment:
Are integer values always encoded as strings? If not, maybe we should have cases that verify that the integer encoded as an integer also works?

Separately, single quoted strings are valid JavaScript but invalid JSON. If we are testing for a string, shouldn't this be double-quoted?

Reply:
It probably depends on the implementation of the SDK. Since we have to be able to read data produced by 7 different JSON implementations, we should probably err on the side of caution. I added another test.

I originally used double quotes throughout this file but it made the sources pretty hard to read. Added a comment to the top of the file.